New getting started guide (#3704)
FusionAuth/fusionauth-site: astro/src/content/docs/get-started/start-here/step-5.mdx

---
title: "Step 5 - Session Management"
description: Learn how to use FusionAuth by using an example application that uses refresh tokens for session management.
section: get started
subcategory: start here
navOrder: 0
codeRoot: https://raw.githubusercontent.com/fusionauth/fusionauth-example-get-started/main
---

<div className="border border-green-700 px-3 rounded-md">
  <i className="fa-duotone fa-stopwatch fa-2x mr-2 text-green-700 align-middle"/> Estimated time to complete this step: **5 minutes**
</div>

## Sessions

> When using FusionAuth as an OAuth identity provider, user's are logged in using what is known as the Authorization Code Grant. This leverages FusionAuth's OAuth frontend to authenticate users and provide an access token to the application.

Reviewer: There was a mix of plural and singular, moved everything to singular.
```suggestion
When using FusionAuth as an OAuth identity provider, a user is logged in using what is known as the Authorization Code Grant. This leverages FusionAuth's OAuth frontend to authenticate the user and provide an access token to the application.
```

Access tokens are designed to have short lifespans though. In most cases, an access token will only be valid for a few minutes before it expires. It would be brutal to force users to log in every few minutes. Instead, it would be ideal to control how log the user is logged in via a user session.

Luckily, the OAuth standard provides a simple way to manage user sessions. This is done using refresh tokens.

> Refresh tokens are long lived tokens that should never be shared with any other application, user, or system. These are the most secure tokens and therefore most be controlled. In many cases, refresh tokens are stored in cookies in the browser. These cookies must be marked as `HttpOnly` and `Secure`. These two settings ensure that the refresh token cannot be stolen in any way. Ideally, the access token is also marked with the same attributes.

Reviewer: ```suggestion
Refresh tokens are long-lived tokens that should never be shared with any other application, user, or system. These are the most secure tokens and therefore access to them must be strictly controlled. In many cases, refresh tokens are stored in cookies in the browser. These cookies must be marked as `HttpOnly` and `Secure`. These two settings ensure that the refresh token cannot be stolen in any way. Ideally, the access token is also marked with the same attributes.
```

  if (this.configuration.enableRefreshTokens && accessToken.refresh_token) {
    res.cookie(this.configuration.refreshTokenCookieName, accessToken.refresh_token, { httpOnly: true });
  }
}
```

You can see that the access and refresh token are written out as `HttpOnly`. Our example application doesn't run over TLS, therefore we can't set the `Secure` flag. But for production applications, you should use TLS and mark cookies as `Secure`.

Next, let's look at how the refresh token is used when the access token expires. Scroll to the `getUser` function in this file. This function loads the same cookie that was written out above. We are using the `jose` library, which throws an exception when the access token has expired. This exception is handled in the `catch` block like this:

```typescript
try {
  payload = (await jose.jwtVerify(accessToken, this.JWKS, {
    issuer: this.configuration.oauthIssuer,
    audience: this.configuration.applicationId,
  })).payload;
} catch (e) {
  if (e instanceof jose.errors.JWTExpired) {
    // ...
  }
}
```

Inside this `catch` block, we check if the exception is due to an expired JWT. If the JWT is expired, we refresh it using this code:

> ```typescript

Reviewer: If I'm not reading carefully, it's easy to miss that this block of code is what's in the `catch` block. It might be clearer to move this code into a `attemptTokenRefresh` function, and call that from the try block and put a call to it in the `catch` block and not just a "// ...". But if you did that then you'd probably want to move the cookie-writing code into a function so that the `catch` block doesn't have a ton of code in it in earlier places in the guide, where it's just going to be confusing/distracting.

if (!response) {
  return null;
}

// Update the cookies making the assumption that they are both valid since we just got them from FusionAuth
accessToken = response.access_token;
res.cookie(this.configuration.accessTokenCookieName, accessToken, { httpOnly: true });

if (this.configuration.enableRefreshTokens && response.refresh_token) {
  refreshToken = response.refresh_token;
  res.cookie(this.configuration.refreshTokenCookieName, refreshToken, { httpOnly: true });
}

payload = jose.decodeJwt(response.id_token);
res.cookie(this.configuration.idTokenCookieName, JSON.stringify(payload), { httpOnly: false });
```

You can see that we first verify if refresh tokens have been enabled and then load the refresh token from the cookie. The code then uses the FusionAuth Typescript client library to call a standard OAuth API that handles what is known as the `Refresh Grant`. You can read up more about this API and the grant in our [OAuth documentation](/docs/lifecycle/authenticate-users/oauth/).

Once FusionAuth responds with a new set of tokens, which might include a new access token, new refresh token, and new id token, these are all stored in their respective cookies again.

> If at any point the refresh token expires or is deleted, this code will immediately return `null`, which will send the user back to the login page.

Reviewer: ```suggestion
If at any point the refresh token expires or is deleted, this code will immediately return `null`, which will cause the `userHasAccess` and `userLoggedIn` checks to fail, sending the user back to the login page.
```

---
title: "Step 5 - Session Management"
description: Learn how to use FusionAuth by using an example application that uses refresh tokens for session management.
section: get started
subcategory: start here
navOrder: 0
codeRoot: https://raw.githubusercontent.com/fusionauth/fusionauth-example-get-started/main
---

<div className="border border-green-700 px-3 rounded-md">
  <i className="fa-duotone fa-stopwatch fa-2x mr-2 text-green-700 align-middle"/> Estimated time to complete this step: **5 minutes**
</div>

## Sessions

> When using FusionAuth as an OAuth identity provider, a user is logged in using what is known as the Authorization Code Grant. This leverages FusionAuth's OAuth frontend to authenticate the user and provide an access token to the application.

Reviewer 2: ```suggestion
When using FusionAuth as an OAuth identity provider, a user is logged in using what is known as the Authorization Code grant. This leverages FusionAuth's OAuth frontend to authenticate the user and provide an access token to the application.
```

```typescript
logInUser(accessToken: AccessToken, res: Response) {
  res.cookie(this.configuration.accessTokenCookieName, accessToken.access_token, { httpOnly: true });
  res.cookie(this.configuration.idTokenCookieName, JSON.stringify(jose.decodeJwt(accessToken.id_token)), { httpOnly: false });

  if (this.configuration.enableRefreshTokens && accessToken.refresh_token) {
    res.cookie(this.configuration.refreshTokenCookieName, accessToken.refresh_token, { httpOnly: true });
  }
}
```

You can see that the access and refresh token are written out as `HttpOnly`. Our example application doesn't run over TLS, therefore we can't set the `Secure` flag. But for production applications, you should use TLS and mark cookies as `Secure`.

Next, let's look at how the refresh token is used when the access token expires. Scroll to the `getUser` function in this file. This function loads the same cookie that was written out above. We are using the `jose` library, which throws an exception when the access token has expired. This exception is handled in the `catch` block like this:

> ```typescript

Reviewer 2: Again, would love for this to be pulled in using RemoteCode.

try {
  payload = (await jose.jwtVerify(accessToken, this.JWKS, {
    issuer: this.configuration.oauthIssuer,
    audience: this.configuration.applicationId,
  })).payload;
} catch (e) {
  // NOTE: Check if the JWT is expired
  if (e instanceof jose.errors.JWTExpired) {
    // ...
  }
}
```

Inside this `catch` block, we check if the exception is due to an expired JWT. If the JWT is expired, we refresh it using this code:

```typescript

> // Refreshing is disabled, so the user is logged out

Reviewer 2: ```suggestion
// Refreshing is disabled, so the user is logged out after the access token expires
```

if (!this.configuration.enableRefreshTokens) {
  return null;
}

// Load the refresh token from the cookie
let refreshToken = req.cookies[this.configuration.refreshTokenCookieName];
if (!refreshToken) {
  return null;
}

// Try refreshing the token
let response = await this.refreshToken(refreshToken);
if (!response) {
  return null;
}

// Update the cookies making the assumption that they are both valid since we just got them from FusionAuth
accessToken = response.access_token;
res.cookie(this.configuration.accessTokenCookieName, accessToken, { httpOnly: true });

> if (this.configuration.enableRefreshTokens && response.refresh_token) {

Reviewer 2: Why do we need `this.configuration.enableRefreshTokens` here? We just checked that on line 61.

Author: True. This is legacy and I just copy and pasted things around. I'll fix that on the example app and here.

  refreshToken = response.refresh_token;
  res.cookie(this.configuration.refreshTokenCookieName, refreshToken, { httpOnly: true });
}

payload = jose.decodeJwt(response.id_token);
res.cookie(this.configuration.idTokenCookieName, JSON.stringify(payload), { httpOnly: false });
```

You can see that we first verify if refresh tokens have been enabled and then load the refresh token from the cookie. The code then uses the FusionAuth Typescript client library to call a standard OAuth API that handles what is known as the `Refresh Grant`. You can read up more about this API and the grant in our [OAuth documentation](/docs/lifecycle/authenticate-users/oauth/).

Once FusionAuth responds with a new set of tokens, which might include a new access token, new refresh token, and new id token, these are all stored in their respective cookies again.

If at any point the refresh token expires or is deleted, this code will immediately return `null`, which will cause the `userHasAccess` and `userLoggedIn` checks to fail, sending the user back to the login page.

> FusionAuth also provides numerous methods for managing user sessions, including the ability to terminate any session or all sessions.

Reviewer 2: ```suggestion
FusionAuth also provides numerous methods for managing user sessions directly using the API or SDKs, including the ability to terminate any session, sessions that meet certain criteria, or all sessions.
```
